File: src/ReadStorm.Infrastructure/Services/EpubExporter.cs
using System.IO.Compression;
using System.Text;

namespace ReadStorm.Infrastructure.Services;

/// <summary>轻量 EPUB 3 生成器，无外部依赖。</summary>
public static class EpubExporter
{
    public static async Task<string> ExportAsync(
        string workDirectory,
        string title,
        string author,
        int sourceId,
        IReadOnlyList<(string Title, string Content)> chapters,
        CancellationToken cancellationToken = default)
    {
        var workDir = WorkDirectoryManager.NormalizeAndMigrateWorkDirectory(workDirectory);
        var downloadPath = WorkDirectoryManager.GetDownloadsDirectory(workDir);
        Directory.CreateDirectory(downloadPath);

        var safeName = SanitizeFileName($"{title}({author}).epub");
        var outputPath = Path.Combine(downloadPath, safeName);

        await using var fileStream = File.Create(outputPath);
        using var archive = new ZipArchive(fileStream, ZipArchiveMode.Create, leaveOpen: false);

        // mimetype (must be first, uncompressed)
        var mimetypeEntry = archive.CreateEntry("mimetype", CompressionLevel.NoCompression);
        await using (var writer = new StreamWriter(mimetypeEntry.Open(), Encoding.ASCII))
        {
            await writer.WriteAsync("application/epub+zip");
        }

        // META-INF/container.xml
        var containerEntry = archive.CreateEntry("META-INF/container.xml");
        await using (var writer = new StreamWriter(containerEntry.Open(), Encoding.UTF8))
        {
            await writer.WriteAsync(ContainerXml);
        }

        // OEBPS/content.opf
        var opfContent = BuildContentOpf(title, author, chapters);
        var opfEntry = archive.CreateEntry("OEBPS/content.opf");
        await using (var writer = new StreamWriter(opfEntry.Open(), Encoding.UTF8))
        {
            await writer.WriteAsync(opfContent);
        }

        // OEBPS/toc.ncx
        var ncxContent = BuildTocNcx(title, chapters);
        var ncxEntry = archive.CreateEntry("OEBPS/toc.ncx");
        await using (var writer = new StreamWriter(ncxEntry.Open(), Encoding.UTF8))
        {
            await writer.WriteAsync(ncxContent);
        }

        // OEBPS/toc.xhtml (nav)
        var navContent = BuildNavXhtml(chapters);
        var navEntry = archive.CreateEntry("OEBPS/toc.xhtml");
        await using (var writer = new StreamWriter(navEntry.Open(), Encoding.UTF8))
        {
            await writer.WriteAsync(navContent);
        }

        // OEBPS/chapter-N.xhtml
        for (var i = 0; i < chapters.Count; i++)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var chapter = chapters[i];
            var chapterHtml = BuildChapterXhtml(chapter.Title, chapter.Content);
            var chapterEntry = archive.CreateEntry($"OEBPS/chapter-{i + 1}.xhtml");
            await using var writer = new StreamWriter(chapterEntry.Open(), Encoding.UTF8);
            await writer.WriteAsync(chapterHtml);
        }

        return outputPath;
    }

    private const string ContainerXml = """
        <?xml version="1.0" encoding="UTF-8"?>
        <container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">
          <rootfiles>
            <rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/>
          </rootfiles>
        </container>
        """;

    private static string BuildContentOpf(string title, string author, IReadOnlyList<(string Title, string Content)> chapters)
    {
        var sb = new StringBuilder();
        sb.AppendLine("""
            <?xml version="1.0" encoding="UTF-8"?>
            <package xmlns="http://www.idpf.org/2007/opf" unique-identifier="BookId" version="3.0">
              <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
            """);
        sb.AppendLine($"    <dc:identifier id=\"BookId\">urn:uuid:{Guid.NewGuid()}</dc:identifier>");
        sb.AppendLine($"    <dc:title>{EscapeXml(title)}</dc:title>");
        sb.AppendLine($"    <dc:creator>{EscapeXml(author)}</dc:creator>");
        sb.AppendLine($"    <dc:language>zh</dc:language>");
        sb.AppendLine($"    <meta property=\"dcterms:modified\">{DateTimeOffset.UtcNow:yyyy-MM-ddTHH:mm:ssZ}</meta>");
        sb.AppendLine("  </metadata>");
        sb.AppendLine("  <manifest>");
        sb.AppendLine("    <item id=\"ncx\" href=\"toc.ncx\" media-type=\"application/x-dtbncx+xml\"/>");
        sb.AppendLine("    <item id=\"nav\" href=\"toc.xhtml\" media-type=\"application/xhtml+xml\" properties=\"nav\"/>");

        for (var i = 0; i < chapters.Count; i++)
        {
            sb.AppendLine($"    <item id=\"chapter-{i + 1}\" href=\"chapter-{i + 1}.xhtml\" media-type=\"application/xhtml+xml\"/>");
        }

        sb.AppendLine("  </manifest>");
        sb.AppendLine("  <spine toc=\"ncx\">");
        for (var i = 0; i < chapters.Count; i++)
        {
            sb.AppendLine($"    <itemref idref=\"chapter-{i + 1}\"/>");
        }

        sb.AppendLine("  </spine>");
        sb.AppendLine("</package>");
        return sb.ToString();
    }

    private static string BuildTocNcx(string title, IReadOnlyList<(string Title, string Content)> chapters)
    {
        var sb = new StringBuilder();
        sb.AppendLine("""
            <?xml version="1.0" encoding="UTF-8"?>
            <ncx xmlns="http://www.daisy.org/z3986/2005/ncx/" version="2005-1">
              <head>
                <meta name="dtb:depth" content="1"/>
              </head>
            """);
        sb.AppendLine($"  <docTitle><text>{EscapeXml(title)}</text></docTitle>");
        sb.AppendLine("  <navMap>");
        for (var i = 0; i < chapters.Count; i++)
        {
            sb.AppendLine($"    <navPoint id=\"navPoint-{i + 1}\" playOrder=\"{i + 1}\">");
            sb.AppendLine($"      <navLabel><text>{EscapeXml(chapters[i].Title)}</text></navLabel>");
            sb.AppendLine($"      <content src=\"chapter-{i + 1}.xhtml\"/>");
            sb.AppendLine("    </navPoint>");
        }

        sb.AppendLine("  </navMap>");
        sb.AppendLine("</ncx>");
        return sb.ToString();
    }

    private static string BuildNavXhtml(IReadOnlyList<(string Title, string Content)> chapters)
    {
        var sb = new StringBuilder();
        sb.AppendLine("""
            <?xml version="1.0" encoding="UTF-8"?>
            <!DOCTYPE html>
            <html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops">
            <head><title>目录</title></head>
            <body>
              <nav epub:type="toc">
                <h1>目录</h1>
                <ol>
            """);
        for (var i = 0; i < chapters.Count; i++)
        {
            sb.AppendLine($"      <li><a href=\"chapter-{i + 1}.xhtml\">{EscapeXml(chapters[i].Title)}</a></li>");
        }

        sb.AppendLine("    </ol>");
        sb.AppendLine("  </nav>");
        sb.AppendLine("</body>");
        sb.AppendLine("</html>");
        return sb.ToString();
    }

    private static string BuildChapterXhtml(string title, string content)
    {
        var paragraphs = content.Split('\n', StringSplitOptions.RemoveEmptyEntries);
        var sb = new StringBuilder();
        sb.AppendLine("""
            <?xml version="1.0" encoding="UTF-8"?>
            <!DOCTYPE html>
            <html xmlns="http://www.w3.org/1999/xhtml">
            <head>
            """);
        sb.AppendLine($"  <title>{EscapeXml(title)}</title>");
        sb.AppendLine("</head>");
        sb.AppendLine("<body>");
        sb.AppendLine($"  <h2>{EscapeXml(title)}</h2>");
        foreach (var p in paragraphs)
        {
            var trimmed = p.Trim();
            if (!string.IsNullOrWhiteSpace(trimmed))
            {
                sb.AppendLine($"  <p>{EscapeXml(trimmed)}</p>");
            }
        }

        sb.AppendLine("</body>");
        sb.AppendLine("</html>");
        return sb.ToString();
    }

    private static string EscapeXml(string text)
    {
        return text
            .Replace("&", "&amp;")
            .Replace("<", "&lt;")
            .Replace(">", "&gt;")
            .Replace("\"", "&quot;")
            .Replace("'", "&apos;");
    }

    private static string SanitizeFileName(string fileName)
    {
        var invalid = Path.GetInvalidFileNameChars();
        var builder = new StringBuilder(fileName.Length);
        foreach (var ch in fileName)
        {
            builder.Append(invalid.Contains(ch) ? '_' : ch);
        }

        return builder.ToString();
    }
}

File: pureopus/app/src/main/java/com/readstorm/app/infrastructure/services/EpubExporter.kt
package com.readstorm.app.infrastructure.services

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * Lightweight EPUB 3 generator — no external dependencies beyond java.util.zip.
 * Produces a valid EPUB file from title, author, and chapter list.
 */
object EpubExporter {

    suspend fun export(
        workDirectory: String,
        title: String,
        author: String,
        sourceId: Int,
        chapters: List<Pair<String, String>> // (chapterTitle, chapterContent)
    ): String = withContext(Dispatchers.IO) {
        val downloadDir = File(workDirectory, "downloads").also { it.mkdirs() }
        val safeName = "${title}(${author}).epub"
            .replace(Regex("[/\\\\:*?\"<>|]"), "_")
        val outputFile = File(downloadDir, safeName)

        ZipOutputStream(FileOutputStream(outputFile)).use { zip ->
            // 1. mimetype (must be first, uncompressed)
            val mimeEntry = ZipEntry("mimetype")
            mimeEntry.method = ZipEntry.STORED
            val mimeBytes = "application/epub+zip".toByteArray()
            mimeEntry.size = mimeBytes.size.toLong()
            mimeEntry.compressedSize = mimeBytes.size.toLong()
            val crc = java.util.zip.CRC32()
            crc.update(mimeBytes)
            mimeEntry.crc = crc.value
            zip.putNextEntry(mimeEntry)
            zip.write(mimeBytes)
            zip.closeEntry()

            // 2. META-INF/container.xml
            zip.putNextEntry(ZipEntry("META-INF/container.xml"))
            zip.write(containerXml().toByteArray())
            zip.closeEntry()

            // 3. OEBPS/content.opf
            zip.putNextEntry(ZipEntry("OEBPS/content.opf"))
            zip.write(contentOpf(title, author, chapters.size).toByteArray())
            zip.closeEntry()

            // 4. OEBPS/toc.ncx
            zip.putNextEntry(ZipEntry("OEBPS/toc.ncx"))
            zip.write(tocNcx(title, chapters).toByteArray())
            zip.closeEntry()

            // 5. OEBPS/toc.xhtml
            zip.putNextEntry(ZipEntry("OEBPS/toc.xhtml"))
            zip.write(tocXhtml(title, chapters).toByteArray())
            zip.closeEntry()

            // 6. Chapter files
            chapters.forEachIndexed { index, (chTitle, content) ->
                zip.putNextEntry(ZipEntry("OEBPS/chapter-${index + 1}.xhtml"))
                zip.write(chapterXhtml(chTitle, content).toByteArray())
                zip.closeEntry()
            }
        }

        outputFile.absolutePath
    }

    private fun escapeXml(text: String): String =
        text.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&apos;")

    private fun containerXml(): String = """<?xml version="1.0" encoding="UTF-8"?>
<container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">
  <rootfiles>
    <rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/>
  </rootfiles>
</container>"""

    private fun contentOpf(title: String, author: String, chapterCount: Int): String {
        val items = (1..chapterCount).joinToString("\n    ") {
            """<item id="chapter-$it" href="chapter-$it.xhtml" media-type="application/xhtml+xml"/>"""
        }
        val spine = (1..chapterCount).joinToString("\n    ") {
            """<itemref idref="chapter-$it"/>"""
        }
        return """<?xml version="1.0" encoding="UTF-8"?>
<package xmlns="http://www.idpf.org/2007/opf" unique-identifier="BookId" version="3.0">
  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
    <dc:identifier id="BookId">urn:uuid:readstorm-${System.currentTimeMillis()}</dc:identifier>
    <dc:title>${escapeXml(title)}</dc:title>
    <dc:creator>${escapeXml(author)}</dc:creator>
    <dc:language>zh</dc:language>
    <meta property="dcterms:modified">${java.time.Instant.now()}</meta>
  </metadata>
  <manifest>
    <item id="toc" href="toc.xhtml" media-type="application/xhtml+xml" properties="nav"/>
    <item id="ncx" href="toc.ncx" media-type="application/x-dtbncx+xml"/>
    $items
  </manifest>
  <spine toc="ncx">
    $spine
  </spine>
</package>"""
    }

    private fun tocNcx(title: String, chapters: List<Pair<String, String>>): String {
        val navPoints = chapters.mapIndexed { index, (chTitle, _) ->
            """    <navPoint id="navpoint-${index + 1}" playOrder="${index + 1}">
      <navLabel><text>${escapeXml(chTitle)}</text></navLabel>
      <content src="chapter-${index + 1}.xhtml"/>
    </navPoint>"""
        }.joinToString("\n")
        return """<?xml version="1.0" encoding="UTF-8"?>
<ncx xmlns="http://www.daisy.org/z3986/2005/ncx/" version="2005-1">
  <head><meta name="dtb:uid" content="urn:uuid:readstorm"/></head>
  <docTitle><text>${escapeXml(title)}</text></docTitle>
  <navMap>
$navPoints
  </navMap>
</ncx>"""
    }

    private fun tocXhtml(title: String, chapters: List<Pair<String, String>>): String {
        val items = chapters.mapIndexed { index, (chTitle, _) ->
            """      <li><a href="chapter-${index + 1}.xhtml">${escapeXml(chTitle)}</a></li>"""
        }.joinToString("\n")
        return """<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops">
<head><title>${escapeXml(title)}</title></head>
<body>
  <nav epub:type="toc">
    <h1>目录</h1>
    <ol>
$items
    </ol>
  </nav>
</body>
</html>"""
    }

    private fun chapterXhtml(title: String, content: String): String {
        val paragraphs = content.split("\n")
            .filter { it.isNotBlank() }
            .joinToString("\n") { "  <p>${escapeXml(it)}</p>" }
        return """<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>${escapeXml(title)}</title></head>
<body>
  <h2>${escapeXml(title)}</h2>
$paragraphs
</body>
</html>"""
    }
}
